File: python/src/advent_day4.py

class Bingo:
    size = 5
    def __init__(self) -> None:
        self.numbers = []
        self.marked = [[False for i in range(Bingo.size)] for j in range(Bingo.size)]
        self.finished = False

    def print(self):
        for r in self.numbers:
            print(r)

    def mark_number(self, n):
        # if already finished, we don't care
        if self.finished:
            return None
        for i in range(Bingo.size):
            for j in range(Bingo.size):
                if self.numbers[i][j] == n:
                    self.marked[i][j] = True
                    # Check row or column all true
                    if all(self.marked[i]) or \
                        all(self.marked[k][j] for k in range(Bingo.size)):
                        # mark as finished
                        self.finished = True
                        # sum all unmarked
                        s = 0
                        for k in range(Bingo.size):
                            for l in range(Bingo.size):
                                s += self.numbers[k][l] \
                                    if not self.marked[k][l] \
                                    else 0
                        return s
        return None


def read_file(file_name):
    """
    Reads the text file
    """
    with open(file_name, 'r') as file:
        list = [line.strip() for line in file.readlines()]
        return list

def load_bingos(lines):
    bingos = []
    bingo = None
    for line in lines:
        if line == "":
            if bingo is not None:
                bingos.append(bingo)
            bingo = Bingo()
        else:
            bingo.numbers.append([int(x) for x in filter(None, line.split(" "))])
    # if not appended
    if bingos[-1] != bingo:
        bingos.append(bingo)
    return bingos

def play_game(numbers, bingos):
    for n in numbers:
        for b in bingos:
            # if it has won
            res = b.mark_number(n)
            if res:
                print("Bingo!")
                return (n, res)


def play_game_loser(numbers, bingos):
    for n in numbers:
        for b in bingos:
            # if it has won
            res = b.mark_number(n)
            # if all marked, return answer
            if all(b.finished for b in bingos):
                return (n, res)



if __name__ == "__main__":
    lines = read_file("test/advent_day4_test.txt")
    numbers = []
    
    numbers = [int(x) for x in filter(None, lines[0].split(","))]
    bingos = load_bingos(lines[1:])
    for b in bingos:
        b.print()
        print()

File: rust/src/advent_day4.rs
use std::iter::Peekable;
use std::{io::{BufRead, BufReader, Lines}, fs::File};

struct BingoCard {
    numbers:Vec<Vec<u8>>,
    marked:Vec<Vec<bool>>
}

const bingo_size:usize = 5;

impl BingoCard {
    fn from_str(lines: &mut Peekable<Lines<BufReader<File>>>) -> BingoCard {
        let mut bingo = BingoCard{
            numbers: vec![], 
            marked: vec![vec![false; bingo_size], 
                        vec![false; bingo_size], 
                        vec![false; bingo_size], 
                        vec![false; bingo_size], 
                        vec![false; bingo_size]]};
        for i in 0..bingo_size {
            let numbers: Vec<u8> = lines.next().unwrap().unwrap().split_whitespace()
                                            .map(|x| x.parse().unwrap())
                                            .collect();
            bingo.numbers.push(numbers);
        }

        return bingo;
    }

    fn mark_number(&mut self, number: &u8) -> Option<u32> {
        for i in 0..bingo_size{
            for j in 0..bingo_size {
                if self.numbers[i][j] == *number {
                    self.marked[i][j] = true;
                    // look row
                    if self.marked[i].iter().all(|x| *x) {
                        let res = self.numbers[i].iter().fold(0, |a, b| a + b) as u32;
                        return Some(res);
                    }
                    // TODO look column
                }
            }
        }
        return None;
    }


    fn print(&self){
        for l in self.numbers.iter() {
            println!("{:?}", l);
        }
        println!()
    }
}

fn play_bingo(mut bingos: Vec<BingoCard>, mut numbers: Vec<u8>) -> u32 {
    let i = 0;
    for n in numbers.iter_mut() {
        for b in bingos.iter_mut() {
            match b.mark_number(n) {
                None => {},
                Some(result) => {
                    println!("Bingo!");
                    println!("Resulst {}", result);
                    return result * (i as u32);
                }
            }
        }
    }

    return 0;
}



fn main() {
    
    let f = File::open("advent_day4_test.txt").expect("Unable to open");
    let mut lines = BufReader::new(f).lines().peekable();
    let mut numbers: Vec<u8> = vec![];
    let mut bingos: Vec<BingoCard> = vec![];
    let mut i = 0;
    while lines.peek().is_some()  {
        let line = lines.next().unwrap().unwrap();
        // read ballot numbers
         if i == 0 {
            numbers = line.split(",")
                        .map(|x| x.parse().unwrap())
                        .collect()
            
        } else {
            // bingo card numbers
            let bingo = BingoCard::from_str(&mut lines);
            bingo.print();
            bingos.push(bingo);
        }
    
        i += 1;
    }
    println!("Here!!");
    let r = play_bingo(bingos, numbers);
    println!("{}", r);
}



#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bingo_from_str(){
        let feed = " 8  2 23  4 24\n"+
                    "21  9 14 16  7\n"+
                    "6 10  3 18  5\n"+
                    "1 12 20 15 19";
        let bingo = BingoCard::from_str(feed);
        assert_eq!()
    }

    #[test]
    fn test_add() {
        assert_eq!(add(1, 2), 3);
    }
}